torch/_lobpcg.py

```diff
@@ -0,0 +1,719 @@
+            bXret[i] = worker.X[:, :k]
+        return bE.reshape(A.shape[:-2] + (k,)), bXret.reshape(A.shape[:-2] + (m, k))
+
+    X = torch.randn((m, n), dtype=dtype, device=device) if X is None else X
+    assert len(X.shape) == 2 and X.shape == (m, n), (X.shape, (m, n))
+
+    worker = LOBPCG(A, B, X, iK, iparams, fparams, bparams, method, tracker)
+    worker.run()
+    return worker.E[:k], worker.X[:, :k]
+
+
+@torch.jit.script
+class LOBPCG(object):
+    """Worker class of LOBPCG methods.
+    """
+
+    def __init__(self,
+                 A,        # type: Optional[Tensor]
+                 B,        # type: Optional[Tensor]
+                 X,        # type: Tensor
+                 iK,       # type: Optional[Tensor]
+                 iparams,  # type: Dict[str, int]
+                 fparams,  # type: Dict[str, float]
+                 bparams,  # type: Dict[str, bool]
+                 method,   # type: str
+                 tracker   # type: Optional[None]
+                 ):
+        # type: (...) -> None
+
+        # constant parameters
+        self.A = A
+        self.B = B
+        self.iK = iK
+        self.iparams = iparams
+        self.fparams = fparams
+        self.bparams = bparams
+        self.method = method
+        self.tracker = tracker
+        m = iparams['m']
+        n = iparams['n']
+
+        # variable parameters
+        self.X = X
+        self.E = torch.zeros((n, ), dtype=X.dtype, device=X.device)
+        self.R = torch.zeros((m, n), dtype=X.dtype, device=X.device)
+        self.S = torch.zeros((m, 3 * n), dtype=X.dtype, device=X.device)
+        self.tvars = {}               # type: Dict[str, Tensor]
+        self.ivars = {'istep': 0}     # type: Dict[str, int]
+        self.fvars = {'_': 0.0}       # type: Dict[str, float]
+        self.bvars = {'_': False}     # type: Dict[str, bool]
+
+    def __str__(self):
+        lines = ['LOPBCG:']
+        lines += ['  iparams={}'.format(self.iparams)]
+        lines += ['  fparams={}'.format(self.fparams)]
+        lines += ['  bparams={}'.format(self.bparams)]
+        lines += ['  ivars={}'.format(self.ivars)]
+        lines += ['  fvars={}'.format(self.fvars)]
+        lines += ['  bvars={}'.format(self.bvars)]
+        lines += ['  tvars={}'.format(self.tvars)]
+        lines += ['  A={}'.format(self.A)]
+        lines += ['  B={}'.format(self.B)]
+        lines += ['  iK={}'.format(self.iK)]
+        lines += ['  X={}'.format(self.X)]
+        lines += ['  E={}'.format(self.E)]
+        r = ''
+        for line in lines:
+            r += line + '\n'
+        return r
+
+    def update(self):
+        """Set and update iteration variables.
+        """
+        if self.ivars['istep'] == 0:
+            X_norm = float(torch.norm(self.X))
+            iX_norm = X_norm ** -1
+            A_norm = float(torch.norm(_utils.matmul(self.A, self.X))) * iX_norm
+            B_norm = float(torch.norm(_utils.matmul(self.B, self.X))) * iX_norm
```

Comment:
Well, here `X_norm` looks like Frobenius norm, but `A_norm` and `B_norm` more like spectral norm... Very confusing.

Reply:
See Sec 4.3 in https://epubs.siam.org/doi/10.1137/17M1129830 for an explanation.

Reply:
Got it, they do not call it as Reyleigh question, but it is! Cool!
But then the name is a bit confusing, because technically `A_norm` and `B_norm` are not exactly norms.